synapse/handlers/saml_handler.py

```diff
@@ -95,6 +101,181 @@ def handle_redirect_request(self, client_redirect_url):
         # this shouldn't happen!
         raise Exception("prepare_for_authenticate didn't return a Location header")
 
+    # User should have only one local session to only one entity
+    # But in case of crash we may have multiple sessions stored in the
+    # local cache. In that case we select the last session and remove
+    # other ones.
+    def _find_session_from_user(self, username):
+        subjects = self._saml_client.users.subjects()
+        sessions = []
+        # Find all sessions for a specific subject
+        for name_id in subjects:
+            # We should always have one source only
+            for source in self._saml_client.users.sources(name_id):
+                info = self._saml_client.users.get_info_from(name_id, source)
+                # If the username is found, append the session info
+                if (
+                    "ava" in info
+                    and "uid" in info["ava"]
+                    and username in info["ava"]["uid"]
+                ):
+                    info["entity"] = source
+                    sessions.append(info)
+
+        sessions.sort(key=lambda i: i["not_on_or_after"])
+        try:
+            # Retrieve last session
+            last_session = sessions.pop()
+            # We should have only one session active for one person
+            # This should never match but removing these staled sessions anyway
+            [self._saml_client.local_logout(session["name_id"]) for session in sessions]
+            return last_session
+        except IndexError:
+            return None
+
+    def _find_mxid_from_name_id(self, name_id):
+        try:
+            attributes = self._saml_client.users.get_identity(name_id)
+            for attribute in attributes:
+                if self._mxid_source_attribute in attribute:
+                    return "@%s:%s" % (
+                        attribute[self._mxid_source_attribute][0],
+                        self._auth_handler.hs.hostname,
+                    )
+        except Exception:
+            pass
+        return None
+
+    async def _logout(self, mxid):
+        # first delete all of the user's devices
+        await self._device_handler.delete_all_devices_for_user(mxid)
+
+        # .. and then delete any access tokens which weren't associated with
+        # devices.
+        await self._auth_handler.delete_access_tokens_for_user(mxid)
+
+    # Example: https://github.com/IdentityPython/pysaml2/blob/master/example/sp-wsgi/sp.py
+    def create_logout_request(self, user, access_token):
```

Comment:
please could you document the parameter types (preferably via type annotations)? Likewise for a bunch of the other new methods.

Reply:
I added extended explanation to all methods inside annotations. Please tell me if it's understandable. I am french, my english is not perfect :)

Reply:
It looks like we're still missing type information for `user` and `access_token` here.

when I say "type" annotations, I mean following the style of PEP484. Writing the types down inside the docstring is ok but we're trying to switch to annotations.

```suggestion
    def create_logout_request(self, user: str, access_token, str) -> bytes:
```